### ISSUE TITLE
returning None from buildProtocol in ReconnectingClientFactory causes exception

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4598]
|<img alt="djk121's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| djk121 reported|
|-|-|
|Trac ID|trac#710|
|Type|enhancement|
|Created|2004-09-11 02:00:21Z|
|Branch|https://github.com/twisted/twisted/tree/connector-buildProtocol-710-2|

Attachments:

* [test_client.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/86e/86e55aec99bfd3868b399afe0c5069bd2a7b7d18/54a4c82af34e8b564c26e01dd20c842df4b224bf.py) (1081 bytes) - added by squiddo on 2011-03-15 06:14:19Z - Test client that shows the same stack trace

<details><summary>Searchable metadata</summary>

```
trac-id__710 710
type__enhancement enhancement
reporter__djk121 djk121
priority__normal normal
milestone__ 
branch__branches_connector_buildProtocol_710_2 branches/connector-buildProtocol-710-2
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1094868021000000 1094868021000000
changetime__1404436172997188 1404436172997188
version__ 
owner__exarkun exarkun
cc__itamarst cc__teratorn cc__djk121
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="djk121's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|djk121 commented|
|-|-|

```
#!html
<pre>
Traceback (most recent call last):
  File "/ita/python-2.2.2/lib/python2.2/site-packages/twisted/internet/default.py", line 526, in 
doSelect
    _logrun(selectable, _drdw, selectable, method, dict)
  File "/ita/python-2.2.2/lib/python2.2/site-packages/twisted/python/log.py", line 65, in 
callWithLogger
    callWithContext({"system": lp}, func, *args, **kw)
  File "/ita/python-2.2.2/lib/python2.2/site-packages/twisted/python/log.py", line 52, in 
callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "/ita/python-2.2.2/lib/python2.2/site-packages/twisted/python/context.py", line 43, in 
callWithContext
    return func(*args,**kw)
--- &lt;exception caught here> ---
  File "/ita/python-2.2.2/lib/python2.2/site-packages/twisted/internet/default.py", line 535, in 
_doReadOrWrite
    why = getattr(selectable, method)()
  File "/ita/python-2.2.2/lib/python2.2/site-packages/twisted/internet/tcp.py", line 434, in doConnect
    self._connectDone()
  File "/ita/python-2.2.2/lib/python2.2/site-packages/twisted/internet/tcp.py", line 439, in 
_connectDone
    self.protocol.makeConnection(self)
exceptions.AttributeError: 'NoneType' object has no attribute 'makeConnection'
Traceback (most recent call last):
  File "server.py", line 239, in ?
    reactor.run()
  File "/ita/python-2.2.2/lib/python2.2/site-packages/twisted/internet/default.py", line 126, in run
    self.mainLoop()
  File "/ita/python-2.2.2/lib/python2.2/site-packages/twisted/internet/default.py", line 137, in 
mainLoop
    self.doIteration(t)
  File "/ita/python-2.2.2/lib/python2.2/site-packages/twisted/internet/default.py", line 526, in 
doSelect
    _logrun(selectable, _drdw, selectable, method, dict)
--- &lt;exception caught here> ---
  File "/ita/python-2.2.2/lib/python2.2/site-packages/twisted/python/log.py", line 65, in 
callWithLogger
    callWithContext({"system": lp}, func, *args, **kw)
  File "/ita/python-2.2.2/lib/python2.2/site-packages/twisted/python/log.py", line 52, in 
callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "/ita/python-2.2.2/lib/python2.2/site-packages/twisted/python/context.py", line 43, in 
callWithContext
    return func(*args,**kw)
  File "/ita/python-2.2.2/lib/python2.2/site-packages/twisted/internet/default.py", line 551, in 
_doReadOrWrite
    selectable.connectionLost(failure.Failure(why))
  File "/ita/python-2.2.2/lib/python2.2/site-packages/twisted/internet/tcp.py", line 447, in 
connectionLost
    Connection.connectionLost(self, reason)
  File "/ita/python-2.2.2/lib/python2.2/site-packages/twisted/internet/tcp.py", line 294, in 
connectionLost
    protocol.connectionLost(reason)
exceptions.AttributeError: 'NoneType' object has no attribute 'connectionLost'
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)|@teratorn commented|
|-|-|

```
#!html
<pre>
Why would you ever return None from buildProtocol? I just glanced at the source 
and this doesn't make any sense to me.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)|@teratorn commented|
|-|-|

```
#!html
<pre>
I'm going to change this to resolved now, because I'll probably forget about it. 

If there is any good reason to not return a Protocol from buildProtocol we can 
reopen it later.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
ServerFactory allows this, as a way of automatically closing connections (we did
this on buildProtocol layer so you can have a /etc/hosts.allow|deny wrapper
factory). 

The bug was reported for a casse where the client connection's only interest was
whether or not the connection succeeded, at which point the connection was
supposed to be closed. Obviously this is still easy to do without this feature,
so this is at most a wish-list item for symmetry.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

Do we know if this is still a problem?  Is it specifically `ReconnectingClientFactory` or does it apply to all client factories?  Can we get a test case for this behavior, at least?

**Comment by twisted-trac** [comment]
|<img alt="squiddo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|squiddo set owner to squiddo<br>squiddo set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="squiddo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|squiddo removed owner<br>squiddo set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [41318]) Branching to 'connector-buildProtocol-710'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [41319]) Test stream client endpoint's behavior when the wrapped buildProtocol returns None.

refs [#4598](https://github.com/twisted/twisted/issues/4598)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [41320]) Fire the `.connect()` `Deferred` when the wrapped factory returns None.

refs [#4598](https://github.com/twisted/twisted/issues/4598)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [41321]) Add a test for a client factory returning None from buildProtocol.

refs [#4598](https://github.com/twisted/twisted/issues/4598)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [41322]) Implement None-handling for the POSIX reactors.

refs [#4598](https://github.com/twisted/twisted/issues/4598)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [41323]) Implement None-handling for IOCP reactor.

Also, clean up the pyflakes error the previous commit introduced.

refs [#4598](https://github.com/twisted/twisted/issues/4598)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

I changed the TCP client implementation to handle None from buildProtocol by closing the connection.  I also had to change endpoints to handle this case because the good TCP client testing tools use endpoints.  I'm not super happy about the latter but I guess the result is an improvement anyway (the old endpoint behavior was random and useless similar to the old connectTCP behavior).

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_17"></a>|@wallrj set owner to @exarkun|
|-|-|

Thanks exarkun,

Notes:
 1. Merges cleanly
 2. Forced build: https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/connector-buildProtocol-710&num_builds=5
    1. A few spurious errors

Points:
 1. [twisted/topfiles/710.feature](twisted/topfiles/710.feature)
    1. Maybe add the full public API path rather than just ClientCreator eg
       twisted.internet.protocol.ClientCreator subclasses may now return None...
    2. Maybe also mention the change to client endpoint behaviour.
 2. [branches/connector-buildProtocol-710/twisted/internet/tcp.py](branches/connector-buildProtocol-710/twisted/internet/tcp.py)
    1. "613 # Make up a protocol" -- the new comment is helpful, but would it
       also be worth adding a comment explaining that buildProtocol may return
       None and why? Or link to the doc that explains the behaviour that these
       lines are providing.
    2. It does seem like a bit of a hack to create a Protocol only to destroy it
       again. Have you got any ideas about how the code could be refactored to
       improve this?
 3. [branches/connector-buildProtocol-710/twisted/internet/test/test_endpoints.py](branches/connector-buildProtocol-710/twisted/internet/test/test_endpoints.py)
    1. wf = endpoints... -- I like it, but I thought you prefered whole word variables ;)
    2. BogusFactory could be moved into the module and shared with between the tests.
 4. [branches/connector-buildProtocol-710/twisted/internet/test/test_tcp.py](branches/connector-buildProtocol-710/twisted/internet/test/test_tcp.py)
    1. "552 stopOnError(self, reactor) " -- can you add a comment hinting at
       what stopOnError does and why it's needed here. Or maybe call it
       stopOnLoggedError. Maybe both.
    2. "554 class NoneFactory(ServerFactory): " -- it probably doesn't make much
       difference, but it seems to me that this should inherit from ClientFactory.
    3. "562 endpoint = self.endpoints.client(reactor, port.getHost()) " -- I was
       confused to see you passing an IPv4Address instance to what I thought was
       TCP4ClientEndpoint.connect, but I see there's a TCPCreator wrapper
       defined in the test module which handles this. Confusing.
    4. "569 "should have failed with NoProtocol."" -- doesn't this strike you as
       ugly? You're re-testing the new endpoint behaviour.
    5. "575 return connectionLost " -- a comment here would be helpful,
       explaining how this links the deferred chain in with
       serverFactory.protocolConnectionLost (I'm not sure I fully understand
       it).
    6. "578 disconnecting.addErrback(log.err) " -- Is this what triggers
       stopOnError? If so, maybe add an explanatory comment.
    7. "581 reactor.stop() " -- why not just add this using addBoth? Wouldn't
       that remove the need for stopOnError?
    8. I must say I found test_buildProtocolReturnsNone difficult to
       follow. Probably just me. I like the way the variables indicate the state
       of the server and client connections, but I think it could do with some
       further comments.
    9. It seems surprising that it doesn't actually make any assertions, but I
       guess the point is that the previous behaviour
       "exceptions.AttributeError: 'NoneType' object has no attribute
       'connectionLost'" would fail this test.
 5. [branches/connector-buildProtocol-710/twisted/internet/test/reactormixins.py](branches/connector-buildProtocol-710/twisted/internet/test/reactormixins.py)
    1. stopOnError -- Maybe add a line to the docstring giving an example of a
       situation where this function is useful.
 6. [branches/connector-buildProtocol-710/twisted/internet/iocpreactor/tcp.py](branches/connector-buildProtocol-710/twisted/internet/iocpreactor/tcp.py)
    1. See comments above about t.i.tcp.BaseClient
 7. [branches/connector-buildProtocol-710/twisted/internet/endpoints.py](branches/connector-buildProtocol-710/twisted/internet/endpoints.py)
    1. buildProtocol -- add an @raises epydoc string explaining when NoProtocol
       is raised.
 8. Documentation
    1. I was about to say add documentation to the Factory interface, but I see
       it's already
       there. https://twistedmatrix.com/documents/current/api/twisted.internet.interfaces.IProtocolFactory.html
    2. Maybe add a short paragraph / short example to the
       https://twistedmatrix.com/documents/current/core/howto/clients.html
       explaining how factory.buildProtocol can return None as a way of
       detecting open ports without going to the trouble of starting and
       stopping a protocol.
    3. Maybe add a short paragraph / short example to
       https://twistedmatrix.com/documents/current/core/howto/servers.html
       explaining that how to implement a connection refusal policy using
       factory.buildProtocol that returns None for certain client
       addresses. (see comment:4).
 9. My 2c
    1. Do you actually like this special case in buildProtocol? I don't think I
       do. Wouldn't it be better to provide a NullProtocol class for these
       usecases instead?
    2. In which case, is it worth considering deprecating and eventually
       removing the "return None" behaviour from Factory and Port?

Phew, this review seems to have taken ages!

So exarkun, please answer the numbered points above or address them to your
satisfaction. And if you are then satisfied with it, please merge. Or submit for
another review. Up to you.

Thanks again.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

> It does seem like a bit of a hack to create a Protocol only to destroy it again. Have you got any ideas about how the code could be refactored to improve this?

Not in any detail, no.  I'm sure a complete rewrite of tcp.py could clean it up pretty nicely (explicit state machine, new state for "disconnecting without protocol", handle the state appropriately).  I hope that happens someday. :)

> "554 class NoneFactory(ServerFactory): " -- it probably doesn't make much difference, but it seems to me that this should inherit from ClientFactory.

Surprisingly, even the client endpoints take a server factory.

> "562 endpoint = self.endpoints.client(reactor, port.getHost()) " -- I was confused to see you passing an IPv4Address instance to what I thought was TCP4ClientEndpoint.connect, but I see there's a TCPCreator wrapper defined in the test module which handles this. Confusing.

Hmm.  Maybe that `endpoints` attribute should be renamed?

> "569 "should have failed with NoProtocol."" -- doesn't this strike you as ugly? You're re-testing the new endpoint behaviour.

Yes, it does.  I don't see how to avoid it without giving up all of the testing utilities that make this test _only_ 40 lines long (just barely longer than fits on a page in my editor).  Also... maybe I'm not that bothered by it: this code is never meant to run in this test, it's only present to provide a better failure message.

> "581 reactor.stop() " -- why not just add this using addBoth? Wouldn't that remove the need for stopOnError?

`stopOnError` is mainly for errors that don't cause the `Deferred` to fire at all.  I tried to clarify this in the `stopOnError` docstring.

> It seems surprising that it doesn't actually make any assertions, but I guess the point is that the previous behaviour "exceptions.AttributeError: 'NoneType' object has no attribute 'connectionLost'" would fail this test.

Indeed. :/  Not my favorite kind of test.  If it didn't involve the reactor then I would try to write `assertIs(None, successResultOf(disconnecting))` and then at least it would be an active "nothing bad should happen" test instead of a passive one... But given the reactor needs to run it's hard to do this.

> buildProtocol -- add an @raises epydoc string explaining when NoProtocol is raised.

Doesn't ever actually raise `NoProtocol`, but expanded the docs to explain this extra behavior.

> Do you actually like this special case in buildProtocol? I don't think I do. Wouldn't it be better to provide a NullProtocol class for these usecases instead?

Nope.  `NullProtocol` (`DisconnectProtocol`?) sounds like a much better idea.

> In which case, is it worth considering deprecating and eventually removing the "return None" behaviour from Factory and Port?

Probably...

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

(In [42848]) Branching to 'connector-buildProtocol-710-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun commented|
|-|-|

> Maybe add a short paragraph / short example to the ​https://twistedmatrix.com/documents/current/core/howto/clients.html explaining how factory.buildProtocol can return None as a way of detecting open ports without going to the trouble of starting and stopping a protocol

Considering you convinced this should be deprecated and removed and replaced with a `NullProtocol` idiom I think I'll skip writing the docs. :)

Other points addressed, merging.

Thanks for the review!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun set status to `closed`|
|-|-|

(In [42850]) Merge connector-buildProtocol-710-2

Author: exarkun
Reviewer: rwall
Fixes: [#4598](https://github.com/twisted/twisted/issues/4598)
Fixes: [#6976](https://github.com/twisted/twisted/issues/6976)

Explicitly handle the case where `None` is returned from `buildProtocol`
when using any of the implementations of either `IReactorTCP.connectTCP`
or `IStreamClientEndpoint.connect` in Twisted.
